// ng-morph/json/classes/json-file-content.js
"use strict";
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONFileContent = void 0;
const jsonc_parser_1 = require("jsonc-parser");
class JSONFileContent {
    constructor(content) {
        this.content = content;
    }
    getContent() {
        return this.content;
    }
    get(jsonPath) {
        if (jsonPath.length === 0) {
            return jsonc_parser_1.getNodeValue(this.jsonAst());
        }
        const node = jsonc_parser_1.findNodeAtLocation(this.jsonAst(), jsonPath);
        return node === undefined ? undefined : jsonc_parser_1.getNodeValue(node);
    }
    modify(jsonPath, value, insertInOrder) {
        let getInsertionIndex;
        if (insertInOrder === undefined) {
            const property = jsonPath.slice(-1)[0];
            getInsertionIndex = (properties) => [...properties, property].sort().findIndex((p) => p === property);
        }
        else if (insertInOrder !== false) {
            getInsertionIndex = insertInOrder;
        }
        const edits = jsonc_parser_1.modify(this.content, jsonPath, value, {
            getInsertionIndex,
            formattingOptions: {
                insertSpaces: true,
                tabSize: 2,
            },
        });
        this.content = jsonc_parser_1.applyEdits(this.content, edits);
        this._jsonAst = undefined;
    }
    remove(jsonPath) {
        if (this.get(jsonPath) !== undefined) {
            this.modify(jsonPath, undefined);
        }
    }
    jsonAst() {
        if (this._jsonAst) {
            return this._jsonAst;
        }
        const errors = [];
        this._jsonAst = jsonc_parser_1.parseTree(this.content, errors);
        if (errors.length) {
            const { error, offset } = errors[0];
            throw new Error(`Failed to parse as JSON AST Object. ${jsonc_parser_1.printParseErrorCode(error)} at location: ${offset}.`);
        }
        return this._jsonAst;
    }
}
exports.JSONFileContent = JSONFileContent;
